### PR TITLE
WIP: Complete reimplementation of getindex and setindex

### PR DESCRIPTION
**Pull request by meggart**
Since there were quite some performance and correctness bugs related to batchgetindex reported that proved to be quite hard to patch, I started this branch to rewrite the getindex/setindex implementation, in particular #135 #136 #137 and #138 . This makes the "strange indexing" schemes like Bool arrays, Arrays of CartesianIndex and their mixtures with integers and ranges much resolve within the normal indexing paths. 

This seems to improve type stability, performance and correctness for the reported cases.

### CODE REVIEW

**Comment by rafaqz** [comment]
Sorry @meggart this might need a rebase after the TestTypes PR was merged

File: README.md [hunk 1]
```diff
@@ -6,12 +6,13 @@
...
 [![Codecov](https://codecov.io/gh/meggart/DiskArrays.jl/branch/main/graph/badge.svg)](https://codecov.io/gh/meggart/DiskArrays.jl/tree/main)
 
-This package is an attempt to collect utilities for working with n-dimensional array-like data
-structures that do not have considerable overhead for single read operations. Most important
-examples are arrays that represent data on hard disk that are accessed through a C
-library or that are compressed in chunks. It can be inadvisable to make these arrays a subtype
-of `AbstractArray` many functions working with AbstractArrays assume fast random access into single
-values (including basic things like `getindex`, `show`, `reduce`, etc...). Currently supported features are:
+This package provides a collection of utilities for working with n-dimensional array-like data
+structures that do not have considerable overhead for single read operations.
```
Comment (felixcremer) [inline README.md:10]:
```suggestion
structures that do have considerable overhead for single read operations. 
```

File: README.md [hunk 2]
```diff
@@ -6,12 +6,13 @@
...
-structures that do not have considerable overhead for single read operations. Most important
-examples are arrays that represent data on hard disk that are accessed through a C
-library or that are compressed in chunks. It can be inadvisable to make these arrays a subtype
-of `AbstractArray` many functions working with AbstractArrays assume fast random access into single
-values (including basic things like `getindex`, `show`, `reduce`, etc...). Currently supported features are:
+This package provides a collection of utilities for working with n-dimensional array-like data
+structures that do not have considerable overhead for single read operations. 
+Most important examples are arrays that represent data on hard disk that are accessed through a C
+library or that are compressed in chunks. 
+It can be inadvisable to make these arrays a subtype of `AbstractArray` many functions working with AbstractArrays assume fast random access into single values (including basic things like `getindex`, `show`, `reduce`, etc...).
```
Comment (felixcremer) [inline README.md:13]: Isn't this in contradiction to DiskArrays itself being a subtype of AbstractArray?

Comment (rafaqz) [reply README.md:13]: Maybe it should say "direct subtype"

**Comment by meggart** [comment]
Thanks to everyone helping with this and for all the comments. For now I would really like to merge this, even if we delay the release of a new DiskArrays version for 1-2 weeks. It would just be much easier to test downstream packages on this and to fix unrelated bugs without running into merge conflicts. @rafaqz @felixcremer any objections?

**Comment by meggart** [comment]
> are all compacts bounds up to date? I tried to tested it with the latest YAXArrays main and Zarr and NetCDF... and I got a lot of ERROR: Unsatisfiable requirements detected for package.

Yes because this is marked as breaking we could even tag a release after merging without breaking the downstream packages, since we would need to increase their compats anyway before any of this gets into user's julia installations.

**Comment by meggart** [comment]
And I have already created branches for NCDatasets, Zarr and NetCDF and there all tests pass so far.

**Pull request merged**